docs: PEM-2364 kubeconfig documentatin (#1425)
spectrocloud/librarium: content/docs/04-clusters/06-cluster-management/11.5-kubeconfig.md

  </WarningBox>

* Is the [Spectro Proxy](/integrations/frp) enabled for the cluster?

Use the tables below to help you identify which project role you need to access the kubeconfig file for a cluster.

<br />

#### Cluster Admin

The following table shows the *Cluster Admin* role or equivalent provides access to both the Kubeconfig file and the Admin Kubeconfig file whether OIDC and the Spectro Proxy are configured or not. If you use a custom Palette resource role instead of the Palette role Cluster Admin, ensure the custom Palette resource role has the *delete* permissions for the resource key *cluster* to access both kubeconfig files for a cluster.

 **Is OIDC Configured?** | **Is Spectro Proxy Enabled?** | **Access to Kubeconfig File** | **Access to Admin Kubeconfig File** |
 --- | --- | --- | --- |
Yes | Yes | ✅ | ✅ |
No | Yes| ✅ | ✅ |
Yes | No | ✅ | ✅ |

> #### Non-Cluster Admin

Author: I'm declining this suggestion because the content applies to custom roles as mentioned in the text.

The table shows that lower-level project roles, such as the *Cluster Editor* or the *Cluster Viewer*, or custom Palette resource roles that do not have the *delete* permissions for the resource key *cluster* may have access to the kubeconfig file.

If a cluster has OIDC and the Spectro Proxy enabled then the kubeconfig file is available. Or, if the cluster has OIDC enabled and the Spectro Proxy disabled, the kubeconfig file is available.

 **Is OIDC Configured?** | **Is Spectro Proxy Enabled?** | **Access to Kubeconfig File** | **Access to Admin Kubeconfig File** |
| --- | --- | --- | --- |---|
|Yes | Yes | ✅ | ❌
| No | Yes | ❌ | ❌
| Yes | No | ✅ | ❌

# API Access

Palette exposes two API endpoints that you can use to access the kubeconfig file for a host cluster. The endpoints are:

<br />

> * `GET https://api.spectrocloud.com/v1/spectroclusters/{clusterId}/assets/kubeconfig`

Author: That's not in scope.

---
title: "Kubeconfig"
metaTitle: "Kubeconfig"

> metaDescription: "Learn how the to find the kubeconfig file for your cluster and how permissions are managed."

Reviewer: ```suggestion
metaDescription: "Learn how to find the kubeconfig file for your cluster and how permissions are managed."
```

---

import Tabs from 'shared/components/ui/Tabs';
import WarningBox from 'shared/components/WarningBox';
import InfoBox from 'shared/components/InfoBox';

# Overview

A [kubeconfig](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/) file is a configuration file used to access a Kubernetes cluster. It contains information such as the cluster's API server address, authentication credentials, and cluster-specific settings. The kubeconfig file allows you to authenticate and interact with the cluster using the kubectl CLI or other Kubernetes client tools.

The kubeconfig file is crucial in enabling you and other users to issue kubectl commands against the host cluster. It ensures you have the necessary permissions and access to the cluster's resources. Using the kubeconfig file, you can validate your access to the host cluster and perform various operations, such as deploying applications, managing resources, and monitoring the cluster.

Overall, the kubeconfig file serves as a key component in connecting and interacting with a Kubernetes cluster, providing you with the necessary configuration and credentials to access the cluster's API server.

> You can download the kubeconfig file from the cluster details page in Palette. Check out to the [Access Cluster with CLI](/clusters/cluster-management/palette-webctl) guide for steps on how to download your cluster's kubeconfig file and connect to your host cluster with the kubectl CLI.

Reviewer: ```suggestion
You can download the kubeconfig file from the cluster details page in Palette. Check out the [Access Cluster with CLI](/clusters/cluster-management/palette-webctl) guide for steps on how to download your cluster's kubeconfig file and connect to your host cluster with the kubectl CLI.
```

As a rule of thumb, users with the Palette project role [*Cluster Admin*](/user-management/palette-rbac/project-scope-roles-permissions#cluster) can access the kubeconfig file for all clusters in the project. Users with lower-level project roles such as the  *Cluster Editor* or the *Cluster Viewer* may not be able to access the kubeconfig file of the cluster. 

<br />

<InfoBox>

Palette has its own RBAC system that is separate from the [Kubernetes RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/) system in the host cluster. The permissions assigned to you in Palette determine what operations you can perform on the cluster. The permissions assigned to you in the host cluster through the Kubernetes RBAC system determine what operations you can perform inside the cluster and on its resources. Refer to the [Palette Roles and Kubernetes Roles](/clusters/cluster-management/cluster-rbac#paletterolesandkubernetesroles) for additional information.

</InfoBox>

The access you have as a user to the kubeconfig file for a cluster depends on the following factors:

<br />

> * Is OIDC configured for the cluster? OIDC is configured in the Kubernetes pack YAML file of the cluster profile. Refer to the respective Kubernetes distribution [Pack reference page](/integrations) to learn more about OIDC configuration.

Reviewer: ```suggestion
* Is OIDC configured for the cluster? OIDC is configured in the Kubernetes pack YAML file of the cluster profile. Refer to the respective Kubernetes distribution on the [Packs List](/integrations) page to learn more about OIDC configuration.
```

  <br />

  <WarningBox>

>     When enabling OIDC, ensure the parameter `oidc-issuer-url` and the `clientConfig` block are properly configured in the Kubernetes Pack YAML. Properly configuring both parameters, you ensure the kubeconfig file is available for all project users. Otherwise, the kubeconfig file will only be available for Cluster Admins or custom roles that have the *delete* permission for the resource key *cluster*.

Reviewer: ```suggestion
    When enabling OIDC, ensure the parameter `oidc-issuer-url` and the `clientConfig` block are properly configured in the Kubernetes Pack YAML. Properly configuring both parameters ensures the kubeconfig file is available for all project users. Otherwise, the kubeconfig file will only be available for Cluster Admins or custom roles that have the *delete* permission for the resource key *cluster*.
```

  </WarningBox>

* Is the [Spectro Proxy](/integrations/frp) enabled for the cluster?

Use the following tables to help you identify what project role you need to access the kubeconfig file for a cluster. 

> 

Reviewer: ```suggestion

<br />
```

Reviewer: Just a bit more space would be nice before the heading in line 64.

### Cluster Admin

> If a custom Palette resource role is created and used instead of the Palette role *Cluster Admin*. The custom Palette resource role must have the *delete* permissions for the resource key *cluster* to access the kubeconfig file for a cluster.

Reviewer: The first sentence is incomplete. Or is it intended to be joined with the second?

Also, do you want to use 'you'? "If you use a custom Palette resource role..."

| **Project Role** | **OIDC Configured** | **Spectro Proxy Enabled** | **Access Kubeconfig File** | 
| --- | --- | --- | --- |
| Cluster Admin | ✅ | ✅ | ✅ |
| Cluster Admin | ❌ | ✅ | ✅ |
| Cluster Admin | ✅ | ❌ | ✅ |

> 

Reviewer: It's probably just me, but I'm not real clear on what this table is telling me. Is it telling me what things I can do if I have Non-Cluster Admin permission? If so, why is Non-Cluster Admin listed three times? Should it be Cluster Editor, Cluster Viewer, and a third role? 

Should column headings be 'Configure OIDC', 'Enable Spectro Proxy', 'Access Kubeconfig File'?

### Non-Cluster Admin

*Non-Cluster Admin* refers to lower-level project roles such as the *Cluster Editor* or the *Cluster Viewer*, or custom Palette resource roles that do not have the *delete* permissions for the resource key *cluster*. 

| **Project Role** | **OIDC Configured** | **Spectro Proxy Enabled** | **Access Kubeconfig File** | 
| --- | --- | --- | --- |
| Non-Cluster Admin | ✅ | ✅ | ✅ |
| Non-Cluster Admin | ❌ | ✅ | ❌ |
| Non-Cluster Admin | ✅ | ❌ | ✅ |

# API Access

Palette exposes two API endpoints that you can use to access the kubeconfig file for a cluster. The endpoints are:

<br />

* `GET https://api.spectrocloud.com/v1/spectroclusters/{clusterId}/assets/kubeconfig` - Returns the kubeconfig file for the cluster. The kubeconfig file is returned as a text string. Access to the kubeconfig file is determined by the permissions assigned to you in Palette.  For additional information, refer to the [Kubeconfig Access Permissions](#kubeconfigaccesspermissions) section. 

> * `GET https://api.spectrocloud.com/v1/spectroclusters/{clusterId}/assets/adminkubeconfig` - Returns the admin kubeconfig file for the cluster. The admin kubeconfig file is returned as a text string. Only users with the Palette project role *Cluster Admin* or with a custom Palette resource role with the resource key *cluster* and the *delete* permission can access the admin kubeconfig file for a cluster.

Reviewer: Just a suggestion... what if you remove the dash that kind of hangs out there alone after the endpoint, and make the description that starts "Returns ... " its own paragraph indented to align with ``Get https://....``
